ofrak_core/ofrak/gui/server.py

```diff
@@ -300,63 +329,77 @@ async def get_resource_range(resource_id):
 
     @exceptions_to_http(SerializedError)
     async def unpack_recursively(self, request: Request) -> Response:
         resource = await self._get_resource_for_request(request)
+        script_str = """
+        await $resource.unpack_recursively()"""
+        await self.script_builder.add_action(resource, script_str, ActionType.UNPACK)
         result = await resource.unpack_recursively()
-        response_pjson = await self._serialize_component_result(result)
-        return json_response(response_pjson)
+        return json_response(await self._serialize_component_result(result))
 
     @exceptions_to_http(SerializedError)
     async def pack(self, request: Request) -> Response:
         resource = await self._get_resource_for_request(request)
+        script_str = """
+
+        await $resource.pack()"""
+        await self.script_builder.add_action(resource, script_str, ActionType.PACK)
         result = await resource.pack()
-        response_pjson = await self._serialize_component_result(result)
-        return json_response(response_pjson)
+        return json_response(await self._serialize_component_result(result))
 
     @exceptions_to_http(SerializedError)
     async def pack_recursively(self, request: Request) -> Response:
         resource = await self._get_resource_for_request(request)
+        script_str = """
+        await $resource.pack_recursively()"""
+        await self.script_builder.add_action(resource, script_str, ActionType.PACK)
         result = await resource.pack_recursively()
-        response_pjson = await self._serialize_component_result(result)
-        return json_response(response_pjson)
+        return json_response(await self._serialize_component_result(result))
 
     @exceptions_to_http(SerializedError)
     async def identify(self, request: Request) -> Response:
         resource = await self._get_resource_for_request(request)
+        script_str = """
+        await $resource.identify()"""
+        await self.script_builder.add_action(resource, script_str, ActionType.MOD)
         result = await resource.identify()
-        response_pjson = await self._serialize_component_result(result)
-        return json_response(response_pjson)
+        return json_response(await self._serialize_component_result(result))
 
     @exceptions_to_http(SerializedError)
     async def data_summary(self, request: Request) -> Response:
         resource = cast(Resource, await self._get_resource_for_request(request))
         result = await resource.run(DataSummaryAnalyzer)
-        response_pjson = await self._serialize_component_result(result)
-        return json_response(response_pjson)
+
+        return json_response(await self._serialize_component_result(result))
 
     @exceptions_to_http(SerializedError)
     async def analyze(self, request: Request) -> Response:
         resource = await self._get_resource_for_request(request)
+        script_str = """
```

Comment:
I agree that the way the script strings are currently constructed relies on non-obvious behavior (by requiring a leading newline) and that it would be better to handle that in `ScriptBuilder`. However, in order to remove the leading newline the resulting docstrings end up much worse, in my opinion.

For example, in order to get the alignment correct the docstrings for the script header and `create_root_resource` need to look like:
```
    boilerplate_header: str = (
r"""    from ofrak import *
    from ofrak.core import *

    async def main(ofrak_context: OFRAKContext):
    """
    )
```

```
        script_str = (
f"""        with open("{name}", "rb") as fh:
            resource_data = fh.read()
        root_resource = await ofrak_context.create_root_resource_from_file("{name}")"""

```

Reply:
Multiline strings that are tricky to format can of course still have leading newlines when it helps, redundant newlines shouldn't harm the generated script, but generally it seems like the server code will look a lot nicer if the script builder always adds newlines (reminds me of a difference between C's `printf` and Python's `print`, I often forget I need a trailing newline when I go back to C and once it bit me badly!).

Reply:
@EdwardLarson  As discussed offline, resolving this for MVP and recording it as a future improvement.